lib/src/git.rs

```diff
@@ -1355,6 +1360,135 @@ pub fn is_special_git_remote(remote: &str) -> bool {
+/// new configuration if necessary.
+fn save_git_config(config: &gix::config::File) -> std::io::Result<()> {
+    let mut config_file = File::create(
+        config
+            .meta()
+            .path
+            .as_ref()
+            .expect("Git repository to have a config file"),
+    )?;
+    config.write_to_filter(&mut config_file, |section| section.meta() == config.meta())
+}
+
+fn git_config_branch_section_ids_by_remote(
+    config: &gix::config::File,
+    remote_name: &str,
+) -> Result<Vec<gix::config::file::SectionId>, GitRemoteManagementError> {
+    config
+        .sections_by_name("branch")
+        .into_iter()
+        .flatten()
+        .filter(|section| {
+            let remote_values = section.values("remote").into_iter();
+            let push_remote_values = section.values("pushRemote").into_iter();
+            remote_values
+                .chain(push_remote_values)
+                .any(|branch_remote_name| *branch_remote_name == remote_name.as_bytes())
+        })
+        .map(|section| {
+            if section.value_names().any(|name| {
+                !name.eq_ignore_ascii_case(b"remote") && !name.eq_ignore_ascii_case(b"merge")
+            }) {
+                return Err(GitRemoteManagementError::NonstandardConfiguration(
+                    remote_name.to_owned(),
+                ));
+            }
+            Ok(section.id())
+        })
+        .collect()
+}
+
+fn rename_remote_in_git_branch_config(
+    config: &mut gix::config::File,
+    old_remote_name: &str,
+    new_remote_name: &str,
+) -> Result<(), GitRemoteManagementError> {
+    for id in git_config_branch_section_ids_by_remote(config, old_remote_name)? {
+        config
+            .section_mut_by_id(id)
+            .expect("found section to exist")
+            .set(
+                "remote"
+                    .try_into()
+                    .expect("'remote' to be a valid value name"),
+                BStr::new(new_remote_name),
+            );
+    }
+    Ok(())
+}
+
+fn remove_remote_from_git_config(
+    config: &mut gix::config::File,
+    remote_name: &str,
+) -> Result<(), GitRemoteManagementError> {
+    let section_ids_to_remove = config
+        .sections_by_name("remote")
+        .into_iter()
+        .flatten()
+        .filter(|section| section.header().subsection_name() == Some(BStr::new(remote_name)))
+        .map(|section| {
+            if section.value_names().any(|name| {
+                !name.eq_ignore_ascii_case(b"url") && !name.eq_ignore_ascii_case(b"fetch")
+            }) {
+                return Err(GitRemoteManagementError::NonstandardConfiguration(
+                    remote_name.to_owned(),
+                ));
+            }
+            Ok(section.id())
+        })
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        .chain(git_config_branch_section_ids_by_remote(
+            config,
+            remote_name,
+        )?);
```

Comment:
I think this should be a separate function. `rename_remote()` shouldn't attempt to delete branch configuration, though it appears to be noop.

Reply:
Done.